File: app/windows/HyperWhisper/Models/HyperWhisperCloudCredits.cs
// HYPERWHISPER CLOUD CREDITS MODEL
// Represents the user's credit balance and account status from HyperWhisper Cloud.
//
// CREDIT SYSTEM:
// - 1 credit = $0.001 USD (micro-dollar precision)
// - Default transcription cost: ~5.5 credits per audio minute (Deepgram Nova-3)
// - Trial users: 100 credits (~18 minutes on the default tier) + IP rate limiting
// - Licensed users: Credits purchased via Polar meters (pay-as-you-go)
//
// API ENDPOINT: GET /usage?identifier=<device_id_or_license_key>
//
// RESPONSE FORMAT:
// {
//   "credits_remaining": 150.5,
//   "minutes_remaining": 24,
//   "credits_per_minute": 6.3,
//   "is_licensed": false,
//   "is_anonymous": true,
//   "resets_at": "2024-01-16T00:00:00Z",
//   "customer_id": null,
//   "message": null
// }
//
// ACCOUNT TYPES:
// - Licensed: Has valid Polar license, unlimited usage (billed per use)
// - Trial: Uses device credits, limited balance
// - Anonymous: IP-based rate limiting, daily reset

using System;
using System.Text.Json.Serialization;

namespace HyperWhisper.Models;

/// <summary>
/// Represents HyperWhisper Cloud credit balance and account status.
/// </summary>
public class HyperWhisperCloudCredits
{
    // =========================================================================
    // CREDIT BALANCE
    // =========================================================================

    /// <summary>
    /// Current credit balance.
    /// 1 credit = $0.001 USD (micro-dollar precision).
    /// </summary>
    [JsonPropertyName("credits_remaining")]
    public double CreditsRemaining { get; set; }

    /// <summary>
    /// Estimated minutes of transcription remaining.
    /// Calculated server-side based on typical audio characteristics.
    /// </summary>
    [JsonPropertyName("minutes_remaining")]
    public int MinutesRemaining { get; set; }

    /// <summary>
    /// Current cost per audio minute in credits.
    /// Varies based on backend provider costs.
    /// Default Deepgram Nova-3 pricing is ~5.5 credits/minute.
    /// </summary>
    [JsonPropertyName("credits_per_minute")]
    public double CreditsPerMinute { get; set; }

    // =========================================================================
    // ACCOUNT STATUS
    // =========================================================================

    /// <summary>
    /// Whether the user has a valid Polar license.
    /// Licensed users have pay-as-you-go billing instead of device credits.
    /// </summary>
    [JsonPropertyName("is_licensed")]
    public bool IsLicensed { get; set; }

    /// <summary>
    /// Whether the user is anonymous (IP-based rate limiting).
    /// Anonymous users have daily reset of quota.
    /// </summary>
    [JsonPropertyName("is_anonymous")]
    public bool IsAnonymous { get; set; }

    /// <summary>
    /// When the daily quota resets (anonymous users only).
    /// Null for licensed/trial users.
    /// </summary>
    [JsonPropertyName("resets_at")]
    public DateTime? ResetsAt { get; set; }

    /// <summary>
    /// Polar customer ID (licensed users only).
    /// Useful for account management and support.
    /// </summary>
    [JsonPropertyName("customer_id")]
    public string? CustomerId { get; set; }

    /// <summary>
    /// Optional server message (e.g., promotional info, warnings).
    /// </summary>
    [JsonPropertyName("message")]
    public string? Message { get; set; }

    // =========================================================================
    // COMPUTED PROPERTIES
    // =========================================================================

    /// <summary>
    /// Formatted balance string for display.
    /// Example: "$0.15 remaining (~24 minutes)"
    /// </summary>
    [JsonIgnore]
    public string FormattedBalance
    {
        get
        {
            var dollars = CreditsRemaining / 1000.0;
            return $"${dollars:F2} remaining (~{MinutesRemaining} minutes)";
        }
    }

    /// <summary>
    /// Short formatted balance for compact display.
    /// Example: "~24 min"
    /// </summary>
    [JsonIgnore]
    public string ShortFormattedBalance => $"~{MinutesRemaining} min";

    /// <summary>
    /// Whether credits are completely exhausted.
    /// </summary>
    [JsonIgnore]
    public bool IsExhausted => CreditsRemaining <= 0;

    /// <summary>
    /// Whether credits are running low (less than 10 minutes remaining).
    /// </summary>
    [JsonIgnore]
    public bool IsLow => MinutesRemaining < 10 && MinutesRemaining > 0;

    /// <summary>
    /// Dollar equivalent of current balance.
    /// </summary>
    [JsonIgnore]
    public double DollarBalance => CreditsRemaining / 1000.0;

    /// <summary>
    /// Account type display string.
    /// </summary>
    [JsonIgnore]
    public string AccountType
    {
        get
        {
            if (IsLicensed) return "Licensed";
            if (IsAnonymous) return "Anonymous";
            return "Trial";
        }
    }

    /// <summary>
    /// Formatted reset time for anonymous users.
    /// Example: "Resets at 12:00 AM"
    /// </summary>
    [JsonIgnore]
    public string? FormattedResetTime
    {
        get
        {
            if (ResetsAt == null) return null;
            var local = ResetsAt.Value.ToLocalTime();
            return $"Resets at {local:h:mm tt}";
        }
    }

    // =========================================================================
    // FACTORY METHODS
    // =========================================================================

    /// <summary>
    /// Creates an empty/unknown credits object for error states.
    /// </summary>
    public static HyperWhisperCloudCredits Empty()
    {
        return new HyperWhisperCloudCredits
        {
            CreditsRemaining = 0,
            MinutesRemaining = 0,
            CreditsPerMinute = 6.3, // Default estimate
            IsLicensed = false,
            IsAnonymous = false
        };
    }
}

File: app/macos/hyperwhisper/Models/HyperWhisperCloudCredits.swift
//
//  HyperWhisperCloudCredits.swift
//  hyperwhisper
//
//  HYPERWHISPER CLOUD CREDIT SYSTEM
//  Data models for tracking usage credits in the HyperWhisper Cloud service.
//
//  CREDIT MODEL:
//  - Trial users are tracked by device_id; licensed users by license_key.
//  - Credit balances and per-minute pricing are computed and enforced
//    server-side; this client only renders the values it is handed.
//
//  ARCHITECTURE:
//  - Credit balance is owned server-side.
//  - This model represents the response from the usage endpoint.
//  - Manager class handles fetching and caching credit balance.
//

import Foundation

/// Credit balance response from HyperWhisper Cloud
/// Returned by GET /usage?identifier=<device_id_or_license_key>
struct HyperWhisperCloudCredits: Codable {
    /// Current credit balance (in credits, not dollars)
    let creditsRemaining: Double

    /// Estimated minutes remaining based on credit balance (computed server-side)
    let minutesRemaining: Int

    /// Conversion factor provided by the server indicating credits consumed per minute of audio
    let creditsPerMinute: Double

    /// Whether this is a licensed user (Polar-tracked) or anonymous user (IP-limited)
    let isLicensed: Bool

    /// Whether this is an anonymous user (IP-based rate limiting)
    let isAnonymous: Bool

    /// When the daily limit resets (for anonymous users only)
    let resetsAt: Date?

    /// Polar customer ID (for licensed users only)
    let customerId: String?

    /// Optional message from server (e.g., for new users)
    let message: String?

    /// Coding keys for JSON decoding (snake_case to camelCase conversion)
    enum CodingKeys: String, CodingKey {
        case creditsRemaining = "credits_remaining"
        case minutesRemaining = "minutes_remaining"
        case creditsPerMinute = "credits_per_minute"
        case isLicensed = "is_licensed"
        case isAnonymous = "is_anonymous"
        case resetsAt = "resets_at"
        case customerId = "customer_id"
        case message
    }

    /// Custom decoder to handle ISO 8601 date strings
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        if let decimalCredits = try? container.decode(Double.self, forKey: .creditsRemaining) {
            creditsRemaining = decimalCredits
        } else {
            // Older workers returned whole-number credits; normalize to Double for compatibility
            let wholeCredits = try container.decode(Int.self, forKey: .creditsRemaining)
            creditsRemaining = Double(wholeCredits)
        }
        minutesRemaining = try container.decode(Int.self, forKey: .minutesRemaining)
        creditsPerMinute = try container.decode(Double.self, forKey: .creditsPerMinute)
        isLicensed = try container.decode(Bool.self, forKey: .isLicensed)
        isAnonymous = try container.decode(Bool.self, forKey: .isAnonymous)
        customerId = try container.decodeIfPresent(String.self, forKey: .customerId)
        message = try container.decodeIfPresent(String.self, forKey: .message)

        // Decode date string if present
        if let dateString = try container.decodeIfPresent(String.self, forKey: .resetsAt) {
            let formatter = ISO8601DateFormatter()
            resetsAt = formatter.date(from: dateString)
        } else {
            resetsAt = nil
        }
    }

    /// Formatted string for display in UI
    /// Example: "$0.87 remaining (~87 minutes)"
    var formattedBalance: String {
        let dollars = creditsRemaining / 1000.0
        return String(format: "$%.2f remaining (~%d minutes)", dollars, minutesRemaining)
    }

    /// Whether user has exhausted their credits
    var isExhausted: Bool {
        creditsRemaining <= 0
    }

    /// Whether user has less than 10 minutes remaining (warning threshold)
    var isLow: Bool {
        minutesRemaining < 10 && minutesRemaining > 0
    }
}

/// Error types for HyperWhisper Cloud operations
enum HyperWhisperCloudError: LocalizedError {
    case insufficientCredits(remaining: Int, required: Int)
    /// Transient network failure (no internet, retry exhaustion) — suppressed in Sentry.
    case transientNetwork(String)
    /// Server contract violation (missing HTTPURLResponse, decode fail, unexpected status) — reported to Sentry.
    case invalidResponse(String)
    case serverError(String)

    var errorDescription: String? {
        switch self {
        case .insufficientCredits:
            return "hyperwhisperCloud.error.insufficientCredits".localized
        case .transientNetwork(let message):
            return "hyperwhisperCloud.error.network".localized(arguments: message)
        case .invalidResponse:
            return "hyperwhisperCloud.error.invalidResponse".localized
        case .serverError(let message):
            return "hyperwhisperCloud.error.server".localized(arguments: message)
        }
    }
}

/// Parsed details from a HyperWhisper Cloud 402 response.
///
/// The backend may serialize credit values as decimals. Keep that precision while
/// deciding whether a denial is an expected billing state or a contradictory
/// server response, then down-convert only for the legacy `TranscriptionError`
/// payload.
struct HyperWhisperCloudCreditDenial: Equatable {
    let remaining: Double?
    let required: Double?
    let limit: Double?
    let message: String?

    init(errorJson: [String: Any]?, message: String?) {
        let messageCredits = Self.parseMessageCredits(message)
        self.remaining = Self.number(from: errorJson?["credits_remaining"]) ?? messageCredits.remaining
        self.required = Self.number(from: errorJson?["credits_required"])
        self.limit = Self.number(from: errorJson?["credits_limit"]) ?? messageCredits.limit
        self.message = message
    }

    var remainingForTranscriptionError: Int {
        Self.legacyCreditInt(remaining)
    }

    var requiredForTranscriptionError: Int {
        Self.legacyCreditInt(required)
    }

    var invalidExhaustedBalanceMessage: String? {
        guard let remaining, remaining > 0 else { return nil }
        guard message?.range(of: "exhausted", options: [.caseInsensitive, .diacriticInsensitive]) != nil,
              message?.range(of: "credit", options: [.caseInsensitive, .diacriticInsensitive]) != nil else {
            return nil
        }

        guard let required, required < remaining else { return nil }

        let remainingText = Self.formatCredit(remaining)
        if let limit {
            return "HyperWhisper Cloud returned HTTP 402 for exhausted credits while reporting \(remainingText) of \(Self.formatCredit(limit)) credits remaining"
        }
        return "HyperWhisper Cloud returned HTTP 402 for exhausted credits while reporting \(remainingText) credits remaining"
    }

    private static func legacyCreditInt(_ value: Double?) -> Int {
        guard let value else { return 0 }
        return max(0, Int(value.rounded(.down)))
    }

    private static func number(from value: Any?) -> Double? {
        switch value {
        case let number as NSNumber:
            return number.doubleValue
        case let string as String:
            return Double(string.trimmingCharacters(in: .whitespacesAndNewlines))
        default:
            return nil
        }
    }

    private static func parseMessageCredits(_ message: String?) -> (remaining: Double?, limit: Double?) {
        guard let message else { return (nil, nil) }

        let pattern = #"\byou\s+have\s+([0-9]+(?:\.[0-9]+)?)\s+of\s+([0-9]+(?:\.[0-9]+)?)\s+credits?\s+remaining\b"#
        guard let regex = try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive]) else {
            return (nil, nil)
        }

        let range = NSRange(message.startIndex..<message.endIndex, in: message)
        guard let match = regex.firstMatch(in: message, range: range),
              match.numberOfRanges == 3,
              let remainingRange = Range(match.range(at: 1), in: message),
              let limitRange = Range(match.range(at: 2), in: message) else {
            return (nil, nil)
        }

        return (Double(message[remainingRange]), Double(message[limitRange]))
    }

    private static func formatCredit(_ value: Double) -> String {
        if value.rounded(.towardZero) == value {
            return String(Int(value))
        }
        return String(format: "%.1f", value)
    }
}
